### PR TITLE
UniswapV2Anchor Validator for OCR Aggregators

### PR DESCRIPTION
**Pull request by alexroan**

### CODE REVIEW

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+ *   - Each configured TokenConfig has its own reporter (aggregator) instead of a single reporter. This reporter is responsible for posting prices for a particular token.
+ *   - This contract has an owner (community multisig) with a single privilege: the ability to invalidate a specific reporter.
+ *       Each reporter is invalidated individually, affecting one market at a time rather than every single market. This is because each reporter will consist of a unique decentralised price feed with independent operators.
+ *   - SafeMath functions are removed since this is written for Solidity version ^0.8.0.
+ */
+contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, ConfirmedOwner {
+  using FixedPoint for *;
+
+  /// @notice The number of wei in 1 ETH
+  uint public constant ethBaseUnit = 1e18;
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:34]: why is this needed? You could e.g. write `2 ether` in solidity, see [here](https://docs.soliditylang.org/en/v0.8.3/units-and-global-variables.html) for available units

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 2]
```diff
@@ -0,0 +1,388 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./UniswapConfig.sol";
+import "./UniswapV2OracleLibrary.sol";
+import "../ConfirmedOwner.sol";
+import "../../interfaces/AggregatorValidatorInterface.sol";
+
+struct Observation {
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:9]: I am wondering whether we could improve gas efficiency by packing this into a single slot and using a uint32 for the timestamp. Thoughts?

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+  using FixedPoint for *;
+
+  /// @notice The number of wei in 1 ETH
+  uint public constant ethBaseUnit = 1e18;
+
+  /// @notice A common scaling factor to maintain precision
+  uint public constant expScale = 1e18;
+
+  /// @notice The highest ratio of the new price to the anchor price that will still trigger the price to be updated
+  uint public immutable upperBoundAnchorRatio;
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:40]: So if these ever need to be changed, the contract has to be redeployed and governance has to vote two switch to the new contract?

File: evm-contracts/src/v0.8/dev/Anchors/UniswapConfig.sol [hunk 4]
```diff
@@ -0,0 +1,959 @@
...
+  )
+    public
+    view
+    returns (
+      TokenConfig memory
+    )
+  {
+    require(i < numTokens, "token config not found");
+
+    if (i == 0) return TokenConfig({cToken: cToken00, underlying: underlying00, symbolHash: symbolHash00, baseUnit: baseUnit00, priceSource: priceSource00, fixedPrice: fixedPrice00, uniswapMarket: uniswapMarket00, reporter: reporter00, isUniswapReversed: isUniswapReversed00});
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapConfig.sol:825]: Out of curiosity, does the optimizer turn this into a jump table?

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+  event UniswapWindowUpdated(
+    bytes32 indexed symbolHash,
+    uint oldTimestamp,
+    uint newTimestamp,
+    uint oldPrice,
+    uint newPrice
+  );
+
+  /// @notice The event emitted when reporter invalidates itself
+  event ReporterInvalidated(
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:91]: Can this happen in our case?

Comment (kaleofduty) [reply evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:91]: To clarify, the owner would invalidate the reporter, not the reporter itself. Correct?

File: evm-contracts/src/v0.8/dev/Anchors/UniswapV2OracleLibrary.sol [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    internal
+    view
+    returns (
+      uint32
+    )
+  {
+    return uint32(block.timestamp % 2 ** 32);
+  }
+
+  // produces the cumulative price using counterfactuals to save gas and avoid a call to sync.
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapV2OracleLibrary.sol:65]: what does "counterfactual" mean in this context

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+    uint256,
+    int256 currentAnswer
+  )
+    external
+    override
+    returns (
+      bool valid
+    )
+  {
+    require(currentAnswer >= 0, "current answer cannot be negative");
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:216]: As long as there is no West Texas Intermediate crude on compound... :trollface:

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+    uint256 reporterPrice = uint256(currentAnswer);
+    TokenConfig memory config = getTokenConfigByReporter(msg.sender);
+
+    uint ethPrice = fetchEthPrice();
+    require(config.priceSource == PriceSource.REPORTER, "only reporter prices get posted");
+    uint anchorPrice;
+    if (config.symbolHash == ethHash) {
+      anchorPrice = ethPrice;
+    } else {
+      anchorPrice = fetchAnchorPrice(config.symbolHash, config, ethPrice);
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:226]: Question for myself: what does the ethPrice do here

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+    }
+
+    if (s_reporterInvalidated[msg.sender]) {
+      s_prices[config.symbolHash] = anchorPrice;
+      emit PriceUpdated(config.symbolHash, anchorPrice);
```
Comment (kaleofduty) [inline evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:231]: if the reporter is invalidated, we update the price???

**Review by kaleofduty — COMMENTED** [review]
Left some initial comments and questions.

Have you benchmarked gas costs already?

**Review by kaleofduty — COMMENTED** [review]
some more comments and questions

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol (hunk 3 above)
Comment (alexroan) [reply evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:40]: This has not changed

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+
+  /**
+    * @notice Construct a uniswap anchored view for a set of token configurations
+    * @dev Note that to avoid immature TWAPs, the system must run for at least a single anchorPeriod before using.
+    * @param anchorToleranceMantissa_ The percentage tolerance that the reporter may deviate from the uniswap anchor
+    * @param anchorPeriod_ The minimum amount of time required for the old uniswap price accumulator to be replaced
+    * @param configs The static token configurations which define what prices are supported and how
+    */
+  constructor(
+    uint anchorToleranceMantissa_,
```
[the remark this answers is not in the document]
Comment (alexroan) [reply evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:105]: This is a good point. It's probably better to have every configuration option on pair level.

File: evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+    *  Update new and old observations of lagging window if period elapsed.
+    */
+  function pokeWindowValues(
+    TokenConfig memory config
+  )
+    internal
+    returns (
+      uint,
+      uint,
+      uint
```
[the remark this answers is not in the document]
Comment (alexroan) [reply evm-contracts/src/v0.8/dev/Anchors/UniswapAnchoredView.sol:352]: Removed the names due to compiler warnings, but happy to add them back in.

**Comment by alexroan** [comment]
Closing this and moving to this repo: https://github.com/smartcontractkit/open-oracle/pull/1
